File: area_scan_3d_camera/Basic/ConnectAndCaptureImages/ConnectAndCaptureImages.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can connect to a camera and obtain the 2D image, depth map, and point cloud data.
*/

#include "area_scan_3d_camera/Camera.h"
#include "area_scan_3d_camera/api_util.h"

int main()
{
    // List all available cameras and connect to a camera by the displayed index.
    mmind::eye::Camera camera;
    if (!findAndConnect(camera))
        return -1;

    // Obtain the 2D image and depth map resolutions of the camera.
    mmind::eye::CameraResolutions cameraResolutions;
    showError(camera.getCameraResolutions(cameraResolutions));
    printCameraResolutions(cameraResolutions);

    // Obtain the 2D image.
    mmind::eye::Frame2D frame2D;
    unsigned row = 0;
    unsigned col = 0;
    showError(camera.capture2D(frame2D));
    std::cout << "The size of the 2D image is: " << frame2D.imageSize().width
              << " (width) * " << frame2D.imageSize().height << " (height)." << std::endl;
    try {
        mmind::eye::ColorBGR colorBGR = frame2D.getColorImage().at(row, col);
        std::cout << "The RGB values of the pixel at (" << row << ", " << col
                  << ") is R: " << unsigned(colorBGR.r) << "; G: " << unsigned(colorBGR.g)
                  << "; B: " << unsigned(colorBGR.b) << "." << std::endl;
    } catch (const std::exception& e) {
        std::cout << "Exception: " << e.what() << std::endl;
        camera.disconnect();
        return 0;
    }

    // Obtain the depth map.

    if (!confirmCapture3D()) {
        camera.disconnect();
        return 0;
    }
    mmind::eye::Frame3D frame3D;
    showError(camera.capture3D(frame3D));
    mmind::eye::DepthMap depthMap = frame3D.getDepthMap();

    std::cout << "The size of the depth map is: " << depthMap.width() << " (width) * " << depthMap.height()
              << " (height)." << std::endl;
    try {
        mmind::eye::PointZ depthElem = depthMap.at(row, col);
        std::cout << "The depth value of the pixel at (" << row << ", " << col << ") is " << depthElem.z
                  << " mm." << std::endl;
    } catch (const std::exception& e) {
        std::cout << "Exception: " << e.what() << std::endl;
        camera.disconnect();
        return 0;
    }

    // Obtain the point cloud.
    mmind::eye::PointCloud pointCloud = frame3D.getUntexturedPointCloud();
    std::cout << "The size of the point cloud is" << pointCloud.width()
              << " (width) * " << pointCloud.height() << " (height)." << std::endl;
    try {
        mmind::eye::PointXYZ pointCloudElem = pointCloud.at(row, col);
        std::cout << "The coordinates of the point corresponding to the pixel at (" << row << ", " << col
                  << ") is X: " << pointCloudElem.x << " mm; Y: " << pointCloudElem.y
                  << " mm; Z:" << pointCloudElem.z << " mm." << std::endl;
    } catch (const std::exception& e) {
        std::cout << "Exception: " << e.what() << std::endl;
        camera.disconnect();
        return 0;
    }

    camera.disconnect();
    std::cout << "Disconnected from the camera successfully." << std::endl;
    return 0;
}

File: area_scan_3d_camera/Basic/connect_and_capture_images.py
# With this sample, you can connect to a camera and obtain the 2D image, depth map, and point cloud data.

from mecheye.shared import *
from mecheye.area_scan_3d_camera import *
from mecheye.area_scan_3d_camera_utils import *


class ConnectAndCaptureImages(object):
    def __init__(self):
        self.camera = Camera()

    def connect_and_capture(self):

        # Obtain the 2D image resolution and the depth map resolution of the camera.
        resolution = CameraResolutions()
        show_error(self.camera.get_camera_resolutions(resolution))
        print_camera_resolution(resolution)

        # Obtain the 2D image.
        frame2d = Frame2D()
        show_error(self.camera.capture_2d(frame2d))
        row, col = 222, 222
        color_map = frame2d.get_color_image()
        print("The size of the 2D image is {} (width) * {} (height).".format(
            color_map.width(), color_map.height()))
        rgb = color_map[row * color_map.width() + col]
        print("The RGB values of the pixel at ({},{}) is R:{},G:{},B{}\n".
              format(row, col, rgb.b, rgb.g, rgb.r))

        if not confirm_capture_3d():
            return

        # Obtain the depth map.
        frame3d = Frame3D()
        show_error(self.camera.capture_3d(frame3d))
        depth_map = frame3d.get_depth_map()
        print("The size of the depth map is {} (width) * {} (height).".format(
            depth_map.width(), depth_map.height()))
        depth = depth_map[row * depth_map.width() + col]
        print("The depth value of the pixel at ({},{}) is depth :{}mm\n".
              format(row, col, depth.z))

        # Obtain the point cloud.
        point_cloud = frame3d.get_untextured_point_cloud()
        print("The size of the point cloud is {} (width) * {} (height).".format(
            point_cloud.width(), point_cloud.height()))
        point_xyz = point_cloud[row * depth_map.width() + col]
        print("The coordinates of the point corresponding to the pixel at ({},{}) is X: {}mm , Y: {}mm, Z: {}mm\n".
              format(row, col, point_xyz.x, point_xyz.y, point_xyz.z))

    def main(self):
        # List all available cameras and connect to a camera by the displayed index.
        if find_and_connect(self.camera):
            self.connect_and_capture()
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")


if __name__ == '__main__':
    a = ConnectAndCaptureImages()
    a.main()
